File: apple/MuiOS/MuiOS/MuDatabase.swift
import Foundation
import SQLite3
import MuCore

/// 音樂庫索引 DB（contract/schema.sql v0.2 的 raw-sqlite3 化；≈ Android 的 Room MuDatabase）。
/// 單庫語意：replaceLibrary 全量置換（換資料夾 = 換庫）；
/// trackId 不落庫（ref 輸出時解析——sync-rules §3.2-5）；albums 派生不落庫（§3.2-6）。
/// 與 schema.sql / Room 版的刻意差異（見 schema.sql pins 表註）：
/// - 不建任何 FK（Room 版僅 playlist_items 有 CASCADE，這裡改為顯式 DELETE playlist_items）
/// - pins 不參照 tracks（釘選要在 replaceLibrary 清 tracks 後存活；換庫才清——PinManager.setRoot）
/// 所有方法同步且執行緒安全（內部序列 queue）；呼叫端自行決定在哪個執行緒跑。
final class MuDatabase {

    struct PinRow {
        let trackId: String
        let pinnedAt: Int64
        let state: String
    }

    private var db: OpaquePointer?
    private let queue = DispatchQueue(label: "mu.db")

    init(url: URL) throws {
        try FileManager.default.createDirectory(
            at: url.deletingLastPathComponent(), withIntermediateDirectories: true)
        var handle: OpaquePointer?
        guard sqlite3_open_v2(url.path, &handle,
                              SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE | SQLITE_OPEN_FULLMUTEX,
                              nil) == SQLITE_OK, handle != nil else {
            throw NSError(domain: "MuDatabase", code: 1,
                          userInfo: [NSLocalizedDescriptionKey: "cannot open \(url.path)"])
        }
        db = handle
        try exec(Self.schema)
    }

    deinit {
        if let db { sqlite3_close_v2(db) }
    }

    // MARK: - 引擎狀態

    /// sync_state 的 'root'（目前庫根路徑；無 → nil）。
    func root() -> String? {
        kvGet("root")
    }

    /// iOS 挑選資料夾的 security-scoped bookmark（無/已清 → nil）。
    func bookmark() -> Data? {
        kvGet("root_bookmark").flatMap { $0.isEmpty ? nil : Data(base64Encoded: $0) }
    }

    /// 還原引擎狀態（空 DB → nil；= Android loadEngineState）。
    func loadEngineState() -> EngineState? {
        queue.sync {
            var cursor: [String: String] = [:]
            try? query("SELECT path, rev FROM cursor") { s in
                cursor[text(s, 0)] = text(s, 1)
            }
            var tracks: [String: SyncEngine.IndexedTrack] = [:]
            try? query("SELECT * FROM tracks") { s in tracks[text(s, 1)] = self.rowToIndexedTrack(s) }
            guard !(cursor.isEmpty && tracks.isEmpty) else { return nil }
            var itemsByPath: [String: [(Int, String, Int?)]] = [:]
            try? query("SELECT playlist_id, position, ref, duration_ms FROM playlist_items") { s in
                itemsByPath[text(s, 0), default: []].append((int(s, 1) ?? 0, text(s, 2), int(s, 3)))
            }
            var playlists: [String: SyncEngine.RawPlaylist] = [:]
            try? query("SELECT id, name FROM playlists") { s in
                let items = (itemsByPath[text(s, 0)] ?? [])
                    .sorted { $0.0 < $1.0 }
                    .map { SyncEngine.RawItem(position: $0.0, ref: $0.1, durationMs: $0.2) }
                playlists[text(s, 0)] = SyncEngine.RawPlaylist(name: text(s, 1), items: items)
            }
            var errors: [String: MuCore.Scanner.ScanError] = [:]
            try? query("SELECT path, code FROM scan_errors") { s in
                errors[text(s, 0)] = MuCore.Scanner.ScanError(code: text(s, 1), path: text(s, 0))
            }
            return EngineState(cursor: cursor, tracks: tracks, playlists: playlists, errors: errors)
        }
    }

    /// 全量置換（每輪 sync 後或換庫時；交易內完成；= Android replaceLibrary）。
    /// bookmark：新庫帶入 nil 會沿用既有值（同庫重掃不換書籤）。
    func replaceLibrary(root: String, bookmark: Data?, state: EngineState) {
        queue.sync {
            try? exec("BEGIN IMMEDIATE")
            exec("DELETE FROM tracks")
            exec("DELETE FROM playlist_items")
            exec("DELETE FROM playlists")
            exec("DELETE FROM scan_errors")
            exec("DELETE FROM cursor")
            for it in state.tracks.values {
                run("INSERT OR REPLACE INTO tracks VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)", [
                    .text(it.track.id), .text(it.track.path), .text(it.track.title),
                    .text(it.track.artist), .text(it.track.album), .text(it.track.albumArtist),
                    .text(it.track.albumId), .int(Int64(it.track.disc)), .intOpt(it.track.trackNo),
                    .intOpt(it.track.year), .bool(it.track.compilation),
                    .intOpt(it.track.durationMs), .text(it.track.format),
                    .int(Int64(it.track.sizeBytes)), .null, .text(it.rev),
                    .bool(it.track.tagOk), .bool(it.available),
                ])
            }
            for (path, pl) in state.playlists {
                run("INSERT OR REPLACE INTO playlists VALUES (?,?,?)", [
                    .text(path), .text(path), .text(pl.name),
                ])
                for it in pl.items {
                    run("INSERT OR REPLACE INTO playlist_items VALUES (?,?,?,?)", [
                        .text(path), .int(Int64(it.position)), .text(it.ref),
                        .intOpt(it.durationMs),
                    ])
                }
            }
            for (path, e) in state.errors {
                run("INSERT OR REPLACE INTO scan_errors VALUES (?,?)", [.text(path), .text(e.code)])
            }
            for (path, rev) in state.cursor ?? [:] {
                run("INSERT OR REPLACE INTO cursor VALUES (?,?)", [.text(path), .text(rev)])
            }
            kvSet("root", root)
            kvSet("root_bookmark", bookmark.map { $0.base64EncodedString() } ?? "")
            if sqlite3_exec(db, "COMMIT", nil, nil, nil) != SQLITE_OK {
                exec("ROLLBACK") // 提交失敗（如磁碟滿）不留半開交易
            }
        }
    }

    // MARK: - 釘選（schema pins 表；狀態機見 PinManager）

    func allPins() -> [PinRow] {
        queue.sync {
            var rows: [PinRow] = []
            try? query("SELECT track_id, pinned_at, state FROM pins") { s in
                rows.append(PinRow(trackId: text(s, 0), pinnedAt: int(s, 1).map(Int64.init) ?? 0, state: text(s, 2)))
            }
            return rows
        }
    }

    func upsertPin(trackId: String, state: String) {
        queue.sync {
            run("INSERT OR REPLACE INTO pins VALUES (?,?,?)", [
                .text(trackId), .int(Self.nowMs()), .text(state),
            ])
        }
    }

    func deletePins(_ ids: [String]) {
        guard !ids.isEmpty else { return }
        queue.sync {
            for id in ids {
                run("DELETE FROM pins WHERE track_id = ?", [.text(id)])
            }
        }
    }

    func clearPins() {
        queue.sync { exec("DELETE FROM pins") }
    }

    // MARK: - SQLite 輔助

    private static func nowMs() -> Int64 {
        Int64(Date().timeIntervalSince1970 * 1000)
    }

    private func kvGet(_ key: String) -> String? {
        var out: String?
        queue.sync {
            try? query("SELECT value FROM sync_state WHERE key = ?", [.text(key)]) { s in
                out = text(s, 0)
            }
        }
        return out
    }

    private func kvSet(_ key: String, _ value: String) {
        run("INSERT OR REPLACE INTO sync_state VALUES (?,?)", [.text(key), .text(value)])
    }

    private func rowToIndexedTrack(_ s: OpaquePointer) -> SyncEngine.IndexedTrack {
        // 欄位序 = INSERT 的 tracks VALUES 序
        let t = Track(
            album: text(s, 4), albumArtist: text(s, 5), albumId: text(s, 6),
            artist: text(s, 3), disc: Int(int(s, 7) ?? 1), format: text(s, 12),
            id: text(s, 0), path: text(s, 1), sizeBytes: Int(int(s, 13) ?? 0),
            tagOk: bool(s, 16), title: text(s, 2), trackNo: int(s, 8),
            year: int(s, 9), compilation: bool(s, 10), durationMs: int(s, 11))
        return SyncEngine.IndexedTrack(track: t, rev: text(s, 15), available: bool(s, 17))
    }

    private enum Val {
        case text(String), int(Int64), intOpt(Int?), bool(Bool), null
    }

    private func exec(_ sql: String) {
        sqlite3_exec(db, sql, nil, nil, nil)
    }

    private func run(_ sql: String, _ vals: [Val]) {
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK, let stmt else { return }
        defer { sqlite3_finalize(stmt) }
        for (i, v) in vals.enumerated() {
            let idx = Int32(i + 1)
            switch v {
            case .text(let s): sqlite3_bind_text(stmt, idx, s, -1, Self.transient)
            case .int(let n): sqlite3_bind_int64(stmt, idx, n)
            case .intOpt(let n):
                if let n { sqlite3_bind_int64(stmt, idx, Int64(n)) } else { sqlite3_bind_null(stmt, idx) }
            case .bool(let b): sqlite3_bind_int64(stmt, idx, b ? Int64(1) : Int64(0))
            case .null: sqlite3_bind_null(stmt, idx)
            }
        }
        sqlite3_step(stmt)
    }

    private func query(_ sql: String, _ vals: [Val] = [], _ row: (OpaquePointer) -> Void) throws {
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK, let stmt else {
            throw NSError(domain: "MuDatabase", code: 2,
                          userInfo: [NSLocalizedDescriptionKey: "prepare failed: \(sql)"])
        }
        defer { sqlite3_finalize(stmt) }
        for (i, v) in vals.enumerated() {
            let idx = Int32(i + 1)
            switch v {
            case .text(let s): sqlite3_bind_text(stmt, idx, s, -1, Self.transient)
            case .int(let n): sqlite3_bind_int64(stmt, idx, n)
            case .intOpt(let n):
                if let n { sqlite3_bind_int64(stmt, idx, Int64(n)) } else { sqlite3_bind_null(stmt, idx) }
            case .bool(let b): sqlite3_bind_int64(stmt, idx, b ? Int64(1) : Int64(0))
            case .null: sqlite3_bind_null(stmt, idx)
            }
        }
        while sqlite3_step(stmt) == SQLITE_ROW { row(stmt) }
    }

    private func text(_ s: OpaquePointer, _ i: Int32) -> String {
        sqlite3_column_text(s, i).map { String(cString: $0) } ?? ""
    }

    private func int(_ s: OpaquePointer, _ i: Int32) -> Int? {
        sqlite3_column_type(s, i) == SQLITE_NULL ? nil : Int(sqlite3_column_int64(s, i))
    }

    private func bool(_ s: OpaquePointer, _ i: Int32) -> Bool {
        sqlite3_column_int64(s, i) != 0
    }

    /// SQLITE_TRANSIENT（bind 立即複製）。
    private static let transient = unsafeBitCast(-1, to: sqlite3_destructor_type.self)

    /// contract/schema.sql v0.2 的鏡像（FK 移除理由見類型註解；albums/play_state/favorites 保留未用）。
    private static let schema = """
        PRAGMA user_version = 1;
        CREATE TABLE IF NOT EXISTS tracks (
          id           TEXT PRIMARY KEY,
          path         TEXT NOT NULL UNIQUE,
          title        TEXT NOT NULL,
          artist       TEXT NOT NULL,
          album        TEXT NOT NULL,
          album_artist TEXT NOT NULL,
          album_id     TEXT NOT NULL,
          disc         INTEGER NOT NULL DEFAULT 1,
          track_no     INTEGER,
          year         INTEGER,
          compilation  INTEGER NOT NULL DEFAULT 0,
          duration_ms  INTEGER,
          format       TEXT NOT NULL,
          size_bytes   INTEGER NOT NULL,
          bitrate_kbps INTEGER,
          rev          TEXT NOT NULL,
          tag_ok       INTEGER NOT NULL DEFAULT 0,
          available    INTEGER NOT NULL DEFAULT 1
        );
        CREATE INDEX IF NOT EXISTS idx_tracks_album ON tracks(album_id, disc, track_no, path);
        CREATE INDEX IF NOT EXISTS idx_tracks_artist ON tracks(artist);
        CREATE TABLE IF NOT EXISTS albums (
          id           TEXT PRIMARY KEY,
          name         TEXT NOT NULL,
          album_artist TEXT NOT NULL,
          year         INTEGER,
          compilation  INTEGER NOT NULL DEFAULT 0,
          art_track_id TEXT
        );
        CREATE TABLE IF NOT EXISTS playlists (
          id         TEXT PRIMARY KEY,
          path       TEXT NOT NULL UNIQUE,
          name       TEXT NOT NULL
        );
        CREATE TABLE IF NOT EXISTS playlist_items (
          playlist_id TEXT NOT NULL,
          position    INTEGER NOT NULL,
          ref         TEXT NOT NULL,
          duration_ms INTEGER,
          PRIMARY KEY (playlist_id, position)
        );
        CREATE TABLE IF NOT EXISTS scan_errors (
          path TEXT PRIMARY KEY,
          code TEXT NOT NULL
        );
        CREATE TABLE IF NOT EXISTS cursor (
          path TEXT PRIMARY KEY,
          rev  TEXT NOT NULL
        );
        CREATE TABLE IF NOT EXISTS pins (
          track_id  TEXT PRIMARY KEY,
          pinned_at INTEGER NOT NULL,
          state     TEXT NOT NULL DEFAULT 'wanted'
        );
        CREATE TABLE IF NOT EXISTS play_state (
          id           INTEGER PRIMARY KEY CHECK (id = 1),
          track_id     TEXT,
          position_ms  INTEGER NOT NULL DEFAULT 0,
          updated_at   INTEGER NOT NULL
        );
        CREATE TABLE IF NOT EXISTS favorites (
          track_id     TEXT PRIMARY KEY,
          favorited_at INTEGER NOT NULL
        );
        CREATE TABLE IF NOT EXISTS sync_state (
          key   TEXT PRIMARY KEY,
          value TEXT NOT NULL
        );
        """
}

File: android/app/src/main/java/music/mu/android/db/MuDatabase.kt
package music.mu.android.db

import android.content.Context
import androidx.room.Dao
import androidx.room.Database
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.PrimaryKey
import androidx.room.Query
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.room.Transaction
import mu.core.EngineState
import mu.core.Scanner
import mu.core.SyncEngine

/**
 * 音樂庫索引 DB（contract/schema.sql v0.2 的 Room 化）。
 * 單庫語意：replaceLibrary 全量置換（換資料夾 = 換庫）；
 * trackId 不落庫（ref 輸出時解析——sync-rules §3.2-5）；albums 派生不落庫（§3.2-6）。
 */

@Entity(tableName = "tracks")
data class TrackEntity(
    @PrimaryKey val id: String,          // = path（model.md §2.1 慣例）
    val path: String,
    val title: String,
    val artist: String,
    val album: String,
    val albumArtist: String,
    val albumId: String,
    val disc: Int,
    val trackNo: Int?,
    val year: Int?,
    val compilation: Boolean,
    val durationMs: Long?,
    val format: String,
    val sizeBytes: Long,
    val bitrateKbps: Int?,               // v0 恆 null（schema 保留欄位）
    val rev: String,
    val tagOk: Boolean,
    val available: Boolean,
) {
    constructor(it: SyncEngine.IndexedTrack) : this(
        id = it.track.id, path = it.track.path,
        title = it.track.title, artist = it.track.artist,
        album = it.track.album, albumArtist = it.track.albumArtist,
        albumId = it.track.albumId, disc = it.track.disc,
        trackNo = it.track.trackNo, year = it.track.year,
        compilation = it.track.compilation, durationMs = it.track.durationMs,
        format = it.track.format, sizeBytes = it.track.sizeBytes,
        bitrateKbps = null, rev = it.rev, tagOk = it.track.tagOk,
        available = it.available,
    )

    fun toIndexedTrack() = SyncEngine.IndexedTrack(
        Scanner.Track(
            album = album,
            albumArtist = albumArtist, albumId = albumId, artist = artist,
            disc = disc, format = format, id = id, path = path,
            sizeBytes = sizeBytes, tagOk = tagOk, title = title,
            trackNo = trackNo, year = year, compilation = compilation,
            durationMs = durationMs,
        ),
        rev = rev, available = available,
    )
}

@Entity(tableName = "playlists")
data class PlaylistEntity(
    @PrimaryKey val path: String,
    val name: String,
)

@Entity(
    tableName = "playlist_items",
    primaryKeys = ["playlistPath", "position"],
    foreignKeys = [ForeignKey(
        entity = PlaylistEntity::class, parentColumns = ["path"],
        childColumns = ["playlistPath"], onDelete = ForeignKey.CASCADE,
    )],
)
data class PlaylistItemEntity(
    val playlistPath: String,
    val position: Int,
    val ref: String,
    val durationMs: Int?,
)

@Entity(tableName = "scan_errors")
data class ScanErrorEntity(
    @PrimaryKey val path: String,
    val code: String,
)

@Entity(tableName = "cursor")
data class CursorEntity(
    @PrimaryKey val path: String,
    val rev: String,
)

@Entity(tableName = "sync_state")
data class SyncStateEntity(
    @PrimaryKey val key: String,
    val value: String,
)

/** schema.sql pins 表：wanted|downloading|done|failed（PinManager 狀態機）。 */
@Entity(tableName = "pins")
data class PinEntity(
    @PrimaryKey val trackId: String,
    val pinnedAt: Long,
    val state: String,
)

@Dao
interface LibraryDao {

    @Query("SELECT value FROM sync_state WHERE key = 'root'")
    suspend fun root(): String?

    @Query("SELECT * FROM tracks")
    suspend fun allTracks(): List<TrackEntity>

    @Query("SELECT * FROM playlists")
    suspend fun allPlaylists(): List<PlaylistEntity>

    @Query("SELECT * FROM playlist_items")
    suspend fun allPlaylistItems(): List<PlaylistItemEntity>

    @Query("SELECT * FROM scan_errors")
    suspend fun allErrors(): List<ScanErrorEntity>

    @Query("SELECT * FROM cursor")
    suspend fun allCursor(): List<CursorEntity>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertTracks(xs: List<TrackEntity>)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertPlaylists(xs: List<PlaylistEntity>)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertPlaylistItems(xs: List<PlaylistItemEntity>)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertErrors(xs: List<ScanErrorEntity>)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertCursor(xs: List<CursorEntity>)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun upsertSyncState(x: SyncStateEntity)

    @Query("DELETE FROM tracks")
    suspend fun clearTracks()

    @Query("DELETE FROM playlists")   // playlist_items 靠 CASCADE
    suspend fun clearPlaylists()

    @Query("DELETE FROM scan_errors")
    suspend fun clearErrors()

    @Query("DELETE FROM cursor")
    suspend fun clearCursor()

    @Query("SELECT * FROM pins")
    suspend fun allPins(): List<PinEntity>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun upsertPin(x: PinEntity)

    @Query("DELETE FROM pins WHERE trackId IN (:ids)")
    suspend fun deletePins(ids: List<String>)

    @Query("DELETE FROM pins")
    suspend fun clearPins()

    /** 還原引擎狀態（空 DB → null）。 */
    @Transaction
    suspend fun loadEngineState(): EngineState? {
        val cursor = allCursor().associate { it.path to it.rev }
        val tracks = allTracks().associate { it.path to it.toIndexedTrack() }
        if (cursor.isEmpty() && tracks.isEmpty()) return null
        val items = allPlaylistItems().groupBy { it.playlistPath }
        val playlists = allPlaylists().associate { p ->
            p.path to SyncEngine.RawPlaylist(
                p.name,
                (items[p.path] ?: emptyList())
                    .sortedBy { it.position }
                    .map { SyncEngine.RawItem(it.position, it.ref, it.durationMs) },
            )
        }
        val errors = allErrors().associate { it.path to Scanner.ScanError(it.code, it.path) }
        return EngineState(cursor, tracks, playlists, errors)
    }

    /** 全量置換（每輪 sync 後或換庫時；交易內完成）。 */
    @Transaction
    suspend fun replaceLibrary(root: String, state: EngineState) {
        clearTracks(); clearPlaylists(); clearErrors(); clearCursor()
        insertTracks(state.tracks.values.map { TrackEntity(it) })
        insertPlaylists(state.playlists.map { (p, pl) -> PlaylistEntity(p, pl.name) })
        insertPlaylistItems(state.playlists.flatMap { (p, pl) ->
            pl.items.map { PlaylistItemEntity(p, it.position, it.ref, it.durationMs) }
        })
        insertErrors(state.errors.map { (p, e) -> ScanErrorEntity(p, e.code) })
        insertCursor(state.cursor.orEmpty().map { (p, r) -> CursorEntity(p, r) })
        upsertSyncState(SyncStateEntity("root", root))
    }
}

@Database(
    entities = [
        TrackEntity::class, PlaylistEntity::class, PlaylistItemEntity::class,
        ScanErrorEntity::class, CursorEntity::class, SyncStateEntity::class,
        PinEntity::class,
    ],
    version = 2, // v2 = +pins 表（開發期 fallback 破壞性重建）
    exportSchema = false,
)
abstract class MuDatabase : RoomDatabase() {
    abstract fun libraryDao(): LibraryDao

    companion object {
        fun build(context: Context): MuDatabase =
            Room.databaseBuilder(context, MuDatabase::class.java, "mu.db")
                .fallbackToDestructiveMigration() // 開發期 schema 變動即重建（未發布）
                .build()
    }
}
